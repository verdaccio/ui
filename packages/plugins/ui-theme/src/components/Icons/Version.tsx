import React from 'react';

import { SvgIcon } from '../SvgIcon';

type Props = React.ComponentProps<typeof SvgIcon>;

const Version = React.forwardRef(function Version(props: Props, ref: React.Ref<SVGSVGElement>) {
  return (
    <SvgIcon viewBox="0 0 14 16" height={16} width={14} {...props} ref={ref}>
      <path
        fillRule="evenodd"
        d="M13 3H7c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1h6c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1zm-1 8H8V5h4v6zM4 4h1v1H4v6h1v1H4c-.55 0-1-.45-1-1V5c0-.55.45-1 1-1zM1 5h1v1H1v4h1v1H1c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1z"
      />
    </SvgIcon>
  );
});

Version.displayName = 'Version';

export { Version };
