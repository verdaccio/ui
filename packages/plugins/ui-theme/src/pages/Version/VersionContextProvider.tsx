import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

import { Dispatch, RootState } from '../../store/store';
import { DetailContext } from './context';
import getRouterPackageName from './get-route-package-name';

interface Params {
  scope?: string;
  package: string;
  version?: string;
}

const VersionContextProvider: React.FC = ({ children }) => {
  const { version: packageVersion, package: pkgName, scope } = useParams<Params>();
  const { manifest, readme, packageName, hasNotBeenFound } = useSelector(
    (state: RootState) => state.manifest
  );
  const isLoading = useSelector((state: RootState) => state?.loading?.models.manifest);
  const dispatch = useDispatch<Dispatch>();
  useEffect(() => {
    const packageName = getRouterPackageName(pkgName, scope);
    dispatch.manifest.getManifest({ packageName, packageVersion });
  }, [dispatch, packageVersion, pkgName, scope]);

  return (
    <DetailContext.Provider
      value={{
        packageMeta: manifest,
        packageVersion,
        readMe: readme,
        packageName,
        isLoading,
        hasNotBeenFound,
      }}
    >
      {children}
    </DetailContext.Provider>
  );
};

export default VersionContextProvider;
