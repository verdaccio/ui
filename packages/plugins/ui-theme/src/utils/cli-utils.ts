import { SyntheticEvent } from 'react';

export const copyToClipBoardUtility =
  (str: string): ((e: SyntheticEvent<HTMLElement>) => void) =>
  (event: SyntheticEvent<HTMLElement>): void => {
    event.preventDefault();

    const node = document.createElement('div');
    node.innerText = str;
    if (document.body) {
      document.body.appendChild(node);

      const range = document.createRange();
      const selection = window.getSelection() as Selection;
      range.selectNodeContents(node);
      selection.removeAllRanges();
      selection.addRange(range);
      document.execCommand('copy');
      document.body.removeChild(node);
    }
  };

export function getCLISetConfigRegistry(
  command: string,
  scope: string,
  registryUrl: string
): string {
  // if there is a scope defined there needs to be a ":" separator between the scope and the registry
  return `${command} ${scope}${scope !== '' ? ':' : ''}registry ${registryUrl}`;
}

export function getCLISetRegistry(command: string, registryUrl: string): string {
  return `${command} --registry ${registryUrl}`;
}

export function getCLIChangePassword(command: string, registryUrl: string): string {
  return `${command} profile set password --registry ${registryUrl}`;
}
