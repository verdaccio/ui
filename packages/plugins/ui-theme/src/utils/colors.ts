export const PRIMARY_COLOR = '#4b5e40';
