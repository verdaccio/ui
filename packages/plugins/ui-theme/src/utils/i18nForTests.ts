import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

// NOTE: not used yet, but should replace the original implementation

// https://react.i18next.com/misc/testing#testing-without-stubbing

i18n.use(initReactI18next).init({
  lng: 'en',
  fallbackLng: 'en',

  // have a common namespace used around the full app
  ns: ['translations'],
  defaultNS: 'translations',

  debug: false,

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  resources: { en: { translations: {} } },
});

export default i18n;
