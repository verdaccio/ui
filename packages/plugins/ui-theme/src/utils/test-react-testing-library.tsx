import { StyledEngineProvider } from '@mui/material/styles';
import { render } from '@testing-library/react';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import ThemeProvider from 'verdaccio-ui/design-tokens/ThemeProvider';
import AppConfigurationProvider from 'verdaccio-ui/providers/config';

import i18nConfig from '../i18n/config';

const renderWithStore = (ui, store) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <AppConfigurationProvider>
          <StyledEngineProvider injectFirst={true}>
            <ThemeProvider>
              <I18nextProvider i18n={i18nConfig}>{children}</I18nextProvider>
            </ThemeProvider>
          </StyledEngineProvider>
        </AppConfigurationProvider>
      </Provider>
    ),
  });

const customRender = (node: React.ReactElement, ...options: any) => {
  return render(
    <AppConfigurationProvider>
      <StyledEngineProvider injectFirst={true}>
        <ThemeProvider>
          <I18nextProvider i18n={i18nConfig}>{node}</I18nextProvider>
        </ThemeProvider>
      </StyledEngineProvider>
    </AppConfigurationProvider>,
    ...options
  );
};

export * from '@testing-library/react';
// FIXME: rename all references with customRemder
export { customRender as render };
export { customRender, renderWithStore };
