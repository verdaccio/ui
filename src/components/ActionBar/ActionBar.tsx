import React from 'react';

import { isURL } from 'verdaccio-ui/utils/url';

import { DetailContext } from '../../pages/Version';
import Box from '../Box';

import ActionBarAction, { ActionBarActionProps } from './ActionBarAction';

/* eslint-disable verdaccio/jsx-spread */
const ActionBar: React.FC = () => {
  const detailContext = React.useContext(DetailContext);

  const { packageMeta } = detailContext;

  if (!packageMeta?.latest) {
    return null;
  }

  const { homepage, bugs, dist } = packageMeta.latest;

  const actions: Array<ActionBarActionProps> = [];

  if (homepage && isURL(homepage)) {
    actions.push({ type: 'VISIT_HOMEPAGE', link: homepage });
  }

  if (bugs?.url && isURL(bugs.url)) {
    actions.push({ type: 'OPEN_AN_ISSUE', link: bugs.url });
  }

  if (dist?.tarball && isURL(dist.tarball)) {
    actions.push({ type: 'DOWNLOAD_TARBALL', link: dist.tarball });
  }

  return (
    <Box alignItems="center" display="flex" marginBottom="8px">
      {actions.map(action => (
        <ActionBarAction key={action.link} {...action} />
      ))}
    </Box>
  );
};

export default ActionBar;
