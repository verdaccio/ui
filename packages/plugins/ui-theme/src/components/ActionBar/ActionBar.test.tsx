import React from 'react';
import { cleanup, renderWithStore, screen } from 'verdaccio-ui/utils/test-react-testing-library';

import { DetailContext, DetailContextProps } from '../../pages/Version';
import { store } from '../../store/store';
import ActionBar from './ActionBar';

const detailContextValue: DetailContextProps = {
  packageName: 'foo',
  readMe: 'test',
  enableLoading: () => {},
  isLoading: false,
  hasNotBeenFound: false,
  packageMeta: {
    _uplinks: {},
    latest: {
      name: 'verdaccio-ui/local-storage',
      version: '8.0.1-next.1',
      dist: {
        fileCount: 0,
        unpackedSize: 0,
        tarball: 'http://localhost:8080/bootstrap/-/bootstrap-4.3.1.tgz',
      },
      homepage: 'https://verdaccio.org',
      bugs: {
        url: 'https://github.com/verdaccio/monorepo/issues',
      },
    },
  },
};

const ComponentToBeRendered: React.FC<{ contextValue: DetailContextProps }> = ({
  contextValue,
}) => (
  <DetailContext.Provider value={contextValue}>
    <ActionBar />
  </DetailContext.Provider>
);

describe('<ActionBar /> component', () => {
  afterEach(() => {
    cleanup();
  });

  test('should render the component in default state', () => {
    const { container } = renderWithStore(
      <ComponentToBeRendered contextValue={detailContextValue} />,
      store
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('when there is no action bar data', () => {
    const packageMeta = {
      ...detailContextValue.packageMeta,
      latest: {
        ...detailContextValue.packageMeta.latest,
        homepage: undefined,
        bugs: undefined,
        dist: {
          ...detailContextValue.packageMeta.latest.dist,
          tarball: undefined,
        },
      },
    };

    const { container } = renderWithStore(
      <ComponentToBeRendered contextValue={{ ...detailContextValue, packageMeta }} />,
      store
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('when there is a button to download a tarball', () => {
    renderWithStore(<ComponentToBeRendered contextValue={{ ...detailContextValue }} />, store);
    expect(screen.getByLabelText('Download tarball')).toBeTruthy();
  });

  test('when there is a button to open an issue', () => {
    renderWithStore(<ComponentToBeRendered contextValue={{ ...detailContextValue }} />, store);
    expect(screen.getByLabelText('Open an issue')).toBeTruthy();
  });
});
