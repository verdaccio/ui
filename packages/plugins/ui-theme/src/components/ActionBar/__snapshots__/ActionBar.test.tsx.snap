// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionBar /> component should render the component in default state 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 14px;
}

.emotion-1 {
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Helvetica Neue",Arial,sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.75;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: -apple-system,BlinkMacSystemFont,"Helvetica Neue",Arial,sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  text-transform: uppercase;
  min-height: 36px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
  padding: 0;
  min-width: 0;
  width: 40px;
  height: 40px;
  box-shadow: 0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);
  color: rgba(0, 0, 0, 0.87);
  background-color: #e0e0e0;
  background-color: #4b5e40;
  color: #fff;
  margin-right: 10px;
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3:active {
  box-shadow: 0px 7px 8px -4px rgba(0,0,0,0.2),0px 12px 17px 2px rgba(0,0,0,0.14),0px 5px 22px 4px rgba(0,0,0,0.12);
}

.emotion-3:hover {
  background-color: #f5f5f5;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media (hover: none) {
  .emotion-3:hover {
    background-color: #e0e0e0;
  }
}

.emotion-3.Mui-focusVisible {
  box-shadow: 0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);
}

.emotion-3.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  box-shadow: none;
  background-color: rgba(0, 0, 0, 0.12);
}

.emotion-3:hover {
  color: #4b5e40;
  background: #fff;
}

.emotion-4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-5 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div
  class="MuiBox-root emotion-0"
>
  <a
    aria-label="Visit homepage"
    class=""
    data-mui-internal-clone-element="true"
    href="https://verdaccio.org"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      class="MuiTypography-root MuiTypography-subtitle1 emotion-1"
    >
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall emotion-2 emotion-3"
        data-testid="fab"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-4"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root emotion-5"
        />
      </button>
    </span>
  </a>
  <a
    aria-label="Open an issue"
    class=""
    data-mui-internal-clone-element="true"
    href="https://github.com/verdaccio/monorepo/issues"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      class="MuiTypography-root MuiTypography-subtitle1 emotion-1"
    >
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall emotion-2 emotion-3"
        data-testid="fab"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-4"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root emotion-5"
        />
      </button>
    </span>
  </a>
  <button
    aria-label="Download tarball"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall  emotion-2 emotion-3"
    data-mui-internal-clone-element="true"
    data-testid="fab"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-4"
      data-testid="CloudDownloadIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM17 13l-5 5-5-5h3V9h4v4h3z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root emotion-5"
    />
  </button>
</div>
`;

exports[`<ActionBar /> component when there is no action bar data 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 14px;
}

<div
  class="MuiBox-root emotion-0"
/>
`;
